pkg/querier/queryrange/limits.go

```diff
@@ -71,3 +81,66 @@ func (l cacheKeyLimits) GenerateCacheKey(userID string, r queryrange.Request) st
 	// a cache key can't be reused when an interval changes
 	return fmt.Sprintf("%s:%s:%d:%d:%d", userID, r.GetQuery(), r.GetStep(), currentInterval, split)
 }
+
+type seriesLimiter struct {
+	hashes map[uint64]struct{}
+	rw     sync.RWMutex
+	buf    []byte
+
+	maxSeries int
+	next      queryrange.Handler
+}
+
+type seriesLimiterMiddleware int
+
+// newSeriesLimiter creates a new series limiter middleware for use for a single request.
+func newSeriesLimiter(maxSeries int) queryrange.Middleware {
+	return seriesLimiterMiddleware(maxSeries)
+}
+
+// Wrap wraps a global handler and returns a per request limited handler.
+// The handler returned is thread safe.
+func (slm seriesLimiterMiddleware) Wrap(next queryrange.Handler) queryrange.Handler {
+	return &seriesLimiter{
+		hashes:    make(map[uint64]struct{}),
+		maxSeries: int(slm),
+		buf:       make([]byte, 0, 1024),
```

Comment:
Maybe a message here saying it's just used to minimize allocs would be helpful. It took me a while to come to that conclusion.